Merge old jqlang/jq master to the latest

owenthereal opened the pull request:
After the transfer from stedolan/jq to jqlang/jq, some commits are missing that divert from the older master of stedolan/jq. This is to pull them in: https://github.com/jqlang/jq/issues/2594#issuecomment-1566198375

nicowilliams committed:
Try making some GitHub Actions

nicowilliams committed:
GHA: Add OS X, Windows, and Linux w/ scan-build builds

owenthereal committed:
Add link to discord server

Add link to discord server

owenthereal commented:
cc: @stedolan @nicowilliams

stedolan commented:
Something seems to be wrong with the github action definitions here: the windows one failed with an error and the unix ones have hung.

itchyny commented:
The action workflow does not trigger because the runner `ubuntu-18.04` is unavailable since ~2022-12-01~ April 2023 ([ref](https://github.blog/changelog/2022-08-09-github-actions-the-ubuntu-18-04-actions-runner-image-is-being-deprecated-and-will-be-removed-by-12-1-22/)). Let's use `ubuntu-22.04`, or `ubuntu-latest`.

owenthereal committed:
Update Ubuntu to 22.04 for GitHub Actions

owenthereal committed:
Remove if condition for Ubuntu build

owenthereal committed:
Fix MacOS OS on GitHub Actions

owenthereal committed:
Fix python3 package on GitHub Actions

owenthereal committed:
Disalbe valgrind

Getting failed tests: https://github.com/jqlang/jq/actions/runs/5113866588/jobs/9193542140#step:9:341

owenthereal committed:
Format file

owenthereal committed:
Fix macos build

owenthereal commented:
@stedolan @itchyny @nicowilliams I got the Linux builds going with no problems. But the macos build failed with https://github.com/jqlang/jq/actions/runs/5114162555/jobs/9194126677#step:6:3160. Do you have any insight into what goes wrong? The Linux clang builds ran okay, though.

owenthereal committed:
Fix syntax errors of windows build

https://github.com/jqlang/jq/actions/runs/5114162556/workflow

owenthereal committed:
Install windows package with choco

wader commented:
That is a bit strange, error seems to be:
```
posix.c:94:3: error: implicit declaration of function 'onig_end' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
```
There is an issue related to it https://github.com/jqlang/jq/issues/2381 which seems to indicate it has been fixed in newer version of oniguruma. But why did the linux build not complain about it? i would have guessed linux (with gcc?) would complain and fail also on `-Wimplicit-function-declaration`?

wader commented:
Linux build also complains:
```
posix.c:94:3: warning: implicit declaration of function 'onig_end' is invalid in C99 [-Wimplicit-function-declaration]
```
But does not use `-Werror` for some reason. Both linux and macos build uses clang btw.

owenthereal committed:
Pin oniguruma to 5a24a49d710a9e3bb8ff11d12e1eae5a9f9df40c

MacOS build [fails](https://github.com/jqlang/jq/actions/runs/5114162555/jobs/9194126677#step:6:3160) due to

```
posix.c:94:3: error: implicit declaration of function 'onig_end' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
  onig_end();
```

The current `oniguruma` revision
6fa38f4084b448592888ed9ee43c6e90a46b5f5c (15 Mar 2017) lacks the following explicit declaration in src/onigposix.h:

```
ONIG_EXTERN int onig_end P_((void));
```

This was added to oniguruma in revision 5a24a49d710a9e3bb8ff11d12e1eae5a9f9df40c (8 Sep 2017).

Ref: https://github.com/jqlang/jq/issues/2381

owenthereal commented:
@wader:

> There is an issue related to it https://github.com/jqlang/jq/issues/2381 which seems to indicate it has been fixed in newer version of oniguruma

Thank you! It's an issue with oniguruma's missing onig_end function in the header! I pinned oniguruma to the suggested revision 5a24a49d710a9e3bb8ff11d12e1eae5a9f9df40c as indicated in #2381, and the macos builds are now passing! Weirdly, the Linux clang build is warning while the macos clang build is erroring...

wader commented:
@owenthereal Some addition info: i can reproduce on macOS 13.3.1 with clang 14.0.3 (clang-1403.0.22.14.1) and the error happens during `make check` in one of the oniguruma tests.

But i agree that update is probably the best solution, still confused why macOS clang uses -Werror and Linux does not.

wader commented:
BTW there is a PR https://github.com/jqlang/jq/pull/2561 (fixes https://github.com/jqlang/jq/issues/2558) to update oniguruma to fix an issue also.

wader commented:
Got a bit curious why homebrew jq does not fail https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/jq.rb and it's probably because it does not use builtin oniguruma

owenthereal committed:
Revert windows build to use pacman

owenthereal committed:
Don't zip jq.exe

owenthereal committed:
Fixing windows build

owenthereal commented:
> BTW there is a PR https://github.com/jqlang/jq/pull/2561 (fixes https://github.com/jqlang/jq/issues/2558) to update oniguruma to fix an issue also.

I agree that we should pin to a later version of oniguruma if it fixes other issues. Perhaps let's merge https://github.com/jqlang/jq/pull/2561 after the CI as part of this PR is passing so that we have a way to validate the upgrade wouldn't break anything

owenthereal commented:
Could anyone help with the windows build? I exhausted my windows knowledge to fix it properly :)

wader commented:
> > BTW there is a PR #2561 (fixes #2558) to update oniguruma to fix an issue also.
> 
> I agree that we should pin to a later version of oniguruma if it fixes other issues. Perhaps let's merge #2561 after the CI as part of this PR is passing so that we have a way to validate the upgrade wouldn't break anything

👍  yeap sounds resonable if we have to update oniguruma anyway

wader commented:
> Could anyone help with the windows build? I exhausted my windows knowledge to fix it properly :)

Also severely lacking windows knowledge :( based on the error message and that shell is `cmd` maybe remove the `#`-comment and do `..\configure` instead of `../configure`? or possibly run script using sh/bash somehow (via mingw?)?

wader commented on .github/workflows/windows.yml:
```diff
@@ -0,0 +1,67 @@
+        shell: bash
+        run: |
+          git clone https://github.com/jqlang/jq
+          cd jq
```
wonder if you can skip this and relay on actions/checkout step above? and i would guess the rest of the steps will run with current working directory being the checkout repo

owenthereal commented:
> or possibly run script using sh/bash somehow (via mingw?)?

I tried switching the windows build to use cygwin. But still running into an issue related to line break? https://github.com/jqlang/jq/actions/runs/5115175948/jobs/9196173934?pr=2596#step:4:17.

I would appreciate it if anyone could take on fixing the windows build. We could disable it for now, considering the goal of this PR is to pull in missing commits before the transfer. We could fix the windows build in another PR.

wader commented:
Got a bit further by doing:
```yaml
      - name: Build and Test
        shell: bash
        env:
          SHELLOPTS: igncr # ignore \r in shell script
        run: |
          autoreconf -fi
          ./configure --disable-valgrind --with-oniguruma=builtin --disable-shared --enable-static --enable-all-static
          make
          make SUBDIRS= "TESTS=tests/mantest tests/jqtest tests/onigtest tests/base64test" check
          7z a jq-windows.zip jq.exe
```
Now `autoreconf` runs but fails very non-verbosely on libtoolize:
```
autoreconf-2.71: error: libtoolize failed with exit status: 1
```
... digging

BTW about `SHELLOPTS: igncr` i guess we could split into many `run` instead also but feels a bit weird. I tried to play with `.gitattributes` but didn't get it to work. Could it be that this is how yaml multi-line strings work on windows?

wader commented:
I noticed now that the old appveyor.yml uses msys2 instead of cygwin, maybe we should try to preserve that?

wader commented:
Some more progress using msys2 instead. This successfully build and link jq.exe but `jqtest` fails to run, not sure why, maybe some issue with `#!/bin/sh`... continuing

```yaml
name: Windows Build
on:
  push:
    branches:
      - 'master'
      - 'windows-build'
    paths:
      - '**.[chly]'
      - '**/Makefile.am'
      - '.github/workflows/windows.yml'
  pull_request:
    paths:
      - '**.[chly]'
      - '**/Makefile.am'
      - '.github/workflows/windows.yml'
jobs:
  windows:
    runs-on: windows-latest
    env:
      APPVER: '10.0'
      CODESIGN_PKT: 0000000000000000
      INSTALL_DIR: C:\jq
      WINSDKVER: '10.0.22000.0'
      WIXDIR: 'c:\Program Files (x86)\Windows Installer XML v3.5'
    steps:
      - name: Clone repository
        uses: actions/checkout@v3
        with:
          submodules: true
      - uses: msys2/setup-msys2@v2
        with:
          update: true
          install: >-
            base-devel
            git
            clang
            autoconf
            automake
            libtool
      - shell: msys2 {0}
        run: |
          autoreconf -fiv
          ./configure --disable-valgrind --with-oniguruma=builtin --disable-shared --enable-static --enable-all-static
          make
          make SUBDIRS= "TESTS=tests/mantest tests/jqtest tests/onigtest tests/base64test" check
          7z a jq-windows.zip jq.exe
      - name: Upload Artifacts
        uses: actions/upload-artifact@v2
        with:
          name: jq artifacts
          path: |
            jq-windows.zip
            jq.exe
      - name: Upload Test Logs
        if: true
        uses: actions/upload-artifact@v2
        with:
          name: Test logs
          path: test-suite.log

```

wader commented:
With the below CI config and diff it builds and tests (except one) runs successfully.

Some remaining questions:
- Did jq and jq tests file use to support `\r` line endings or how did this work with appveyor?
- Is the binary produced properly static and has proper Windows version compatibility etc?
- `msys2/setup-msys2@v2` action has a `msystem` option that can be `MSYS | MINGW64 | MINGW32 | UCRT64 | CLANG32 | CLANG64 | CLANGARM64`, we currently use default `MINGW64`. Should be something else?
- Have no idea about packaging, signing and upload

`jq.test:1568` fails with:
```
Test #334: '[range(-52;52;1)] as $powers | [$powers[]|pow(2;.)|log2] == $powers' at line number 1568
*** Expected true, but got false for test at line number 1570: [range(-52;52;1)] as $powers | [$powers[]|pow(2;.)|log2] == $powers
```
Haven't looked deeper at it yet. appveyor builds in different env somehow?


```diff
--- a/.gitattributes
+++ b/.gitattributes
@@ -1,2 +1,4 @@
 .gitattributes export-ignore
 .gitignore export-ignore
+*.test eol=lf
+*.jq eol=lf
--- a/tests/jq.test
+++ b/tests/jq.test
@@ -1565,9 +1565,9 @@ try (1%.) catch .
 jq: error: Division by zero? at <top-level>, line 1:

 # Basic numbers tests: integers, powers of two
-[range(-52;52;1)] as $powers | [$powers[]|pow(2;.)|log2] == $powers
-null
-true
+#[range(-52;52;1)] as $powers | [$powers[]|pow(2;.)|log2] == $powers
+#null
+#true

 [range(-99/2;99/2;1)] as $orig | [$orig[]|pow(2;.)|log2] as $back | ($orig|keys)[]|. as $k | (($orig|.[$k])-($back|.[$k]))|if . < 0 then . * -1 else . end|select(.>.00005)
 null
```

```yaml
name: Windows Build
on:
  push:
    branches:
      - 'master'
      - 'windows-build'
    paths:
      - '**.[chly]'
      - '**/Makefile.am'
      - '.github/workflows/windows.yml'
  pull_request:
    paths:
      - '**.[chly]'
      - '**/Makefile.am'
      - '.github/workflows/windows.yml'
  workflow_dispatch:
jobs:
  windows:
    runs-on: windows-latest
    env:
      APPVER: '10.0'
      CODESIGN_PKT: 0000000000000000
      INSTALL_DIR: C:\jq
      WINSDKVER: '10.0.22000.0'
      WIXDIR: 'c:\Program Files (x86)\Windows Installer XML v3.5'
    steps:
      - name: Clone repository
        uses: actions/checkout@v3
        with:
          submodules: true
      - uses: msys2/setup-msys2@v2
        with:
          update: true
          install: >-
            base-devel
            git
            clang
            autoconf
            automake
            libtool
            p7zip
      - shell: msys2 {0}
        run: |
          autoreconf -fiv
          ./configure --disable-valgrind --with-oniguruma=builtin --disable-shared --enable-static --enable-all-static
          make -j4
          make SUBDIRS= "TESTS=tests/mantest tests/jqtest tests/onigtest tests/base64test" check
          7z a jq-windows.zip jq.exe
      - name: Upload Artifacts
        uses: actions/upload-artifact@v2
        with:
          name: jq artifacts
          path: |
            jq-windows.zip
            jq.exe
      - name: Upload Test Logs
        if: true
        uses: actions/upload-artifact@v2
        with:
          name: Test logs
          path: test-suite.log

```

wader commented:
Modified the failing test a bit to get the array, fails with:
```
Test #334: '[range(-52;52;1)] as $powers | [$powers[]|pow(2;.)|log2]' at line number 1568
*** Expected 
[-52,-51,-50,-49,-48,-47,-46,-45,-44,-43,-42,-41,-40,-39,-38,-37,-36,-35,-34,-33,-32,-31,-30,-29,-28,-27,-26,-25,-24,-23,-22,-21,-20,-19,-18,-17,-16,-15,-14,-13,-12,-11,-10,-9,-8,-7,-6,-5,-4,-3,-2,-1,0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51]
, but got
[-52,-51.00000000000001,-50,-49,-48,-47.00000000000001,-46,-45,-44,-43,-42,-41,-40,-39.00000000000001,-38,-37,-36,-35,-34,-33,-32,-31.000000000000004,-30,-29.000000000000004,-28,-27,-26,-25,-24,-23,-22,-21,-20,-19,-18,-17,-16,-15,-14,-13,-12,-11,-10,-9,-8,-7,-6,-5,-4,-3,-2,-1,0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29.000000000000004,30,31.000000000000004,32,33,34,35,36,37,38,39.00000000000001,40,41,42,43,44,45,46,47.00000000000001,48,49,50,51.00000000000001]
for test at line number 1570: [range(-52;52;1)] as $powers | [$powers[]|pow(2;.)|log2]
```
(added new lines for clarity)

owenthereal commented:
Looking at the test log (downloaded from [here](https://github.com/jqlang/jq/actions/runs/5123520821?pr=2596), there are many failures due to line breaks in `tests/jq.test`. Do we change `.gitattributes` as mentioned in https://github.com/jqlang/jq/pull/2596#issuecomment-1568472921? Wouldn't it break tests for other OSes?

owenthereal committed:
Update .gitattriutes for eol on Windows

wader commented:
Not sure if it will break... but i've started to wonder now if not accepting `\r` as whitespace is a bug? (gojq seems to accept it) did it used to work and is a regression? weirdly it seems like `\n\r` do work, i'm confused.

```sh
$ jq -n $(echo "123\n\r")
123
$ jq -n $(echo "123\n")
123
$ jq -n $(echo "123\r")
jq: error: syntax error, unexpected INVALID_CHARACTER, expecting $end (Unix shell quoting issues?) at <top-level>, line 1:

jq: 1 compile error
```

wader commented on .github/workflows/windows.yml:
```diff
@@ -13,6 +13,7 @@ on:
       - '**.[chly]'
       - '**/Makefile.am'
       - '.github/workflows/windows.yml'
+  workflow_dispatch:
```
You can probably skip this if we don't have the need to manually trigger

owenthereal replied:
Addressed in https://github.com/jqlang/jq/pull/2596/commits/488dc8a4308db36aa396dcf8b44a54bfdff833c8

wader commented:
A patch like this seems to fix the `\r` issue but it would be nice to know why it worked before on windows CI?
```diff
diff --git a/src/lexer.l b/src/lexer.l
index 6b9164b..28d281e 100644
--- a/src/lexer.l
+++ b/src/lexer.l
@@ -123,7 +123,7 @@ struct lexer_param;
 ([a-zA-Z_][a-zA-Z_0-9]*::)*[a-zA-Z_][a-zA-Z_0-9]*  { yylval->literal = jv_string(yytext); return IDENT;}
 \.[a-zA-Z_][a-zA-Z_0-9]*  { yylval->literal = jv_string(yytext+1); return FIELD;}

-[ \n\t]+  {}
+[ \r\n\t]+  {}

 . { return INVALID_CHARACTER; }
```
(could possibly also include `\f\v` which seems to be what gojq do)

wader commented:
It would be great if someone could have a look at the `pow`/`log2` issue, i haven't started yet

itchyny commented:
I don't be surprised `log2(pow2(x))` is not exactly equal to `x` on computer due to floating-point number precision. We can just round the result or assert the absolute difference is small enough.

owenthereal committed:
Skip workflow_dispatch for windows build

owenthereal committed:
Clean up builds

wader commented:
> I don't be surprised `log2(pow2(x))` is not exactly equal to `x` on computer due to floating-point number precision. We can just round the result or assert the absolute difference is small enough.

Good idea, let's do that. Maybe a mystery for later to figure what differs from the old build env.

@owenthereal sounds good to you?

wader commented:
@itchyny any ideas about the `\r` issue? just use gitattributes for now or should we fix something? as `\n` and `\r\n` do work (would be weird otherwise, but just `\r` don't) maybe this is limited to the jq `--run-tests` parser code somehow, haven't checked yet.

itchyny commented:
Using `gitattributes` until fixing the lexer sounds good. I think the defaults of `core.autocrlf` are different between AppVeyor and `actions/checkout`.

owenthereal committed:
Use LF line endings for all text

Ref: https://github.com/actions/checkout/issues/135

owenthereal commented:
> I don't be surprised log2(pow2(x)) is not exactly equal to x on computer due to floating-point number precision. We can just round the result or assert the absolute difference is small enough.

Sounds good to me. Does it require some changes to the test and assertion?

> Using gitattributes until fixing the lexer sounds good. I think the defaults of core.autocrlf are different between AppVeyor and actions/checkout.

I found https://github.com/actions/checkout/issues/135. I followed the recommendation to turn on LG line endings for all text: https://github.com/jqlang/jq/pull/2596/commits/93465d5dbf2fed7c8892c8d4124e22c99d47689c

wader commented:
> Sounds good to me. Does it require some changes to the test and assertion?

Yeap something like this should work i think:
```diff
diff --git a/tests/jq.test b/tests/jq.test
index 2d5c36b..9aa435a 100644
--- a/tests/jq.test
+++ b/tests/jq.test
@@ -1565,7 +1565,7 @@ try (1%.) catch .
 jq: error: Division by zero? at <top-level>, line 1:

 # Basic numbers tests: integers, powers of two
-[range(-52;52;1)] as $powers | [$powers[]|pow(2;.)|log2] == $powers
+[range(-52;52;1)] as $powers | [$powers[]|pow(2;.)|log2|round] == $powers
 null
 true
```
You can run the tests with `jq --run-tests < tests/jq.test`

owenthereal committed:
Fix test that has rounding error

https://github.com/jqlang/jq/pull/2596#issuecomment-1568498993

owenthereal commented:
Weeee....CI is passing finally! Thank you all for the help! Would you happen to have any more feedback? If not, let's merge. At the moment, the CI only compiles & runs tests. A follow-up PR is needed to publish releases to GH Releases.

owenthereal committed:
Enable CI for all

wader commented:
No worries! happy to help and i learned some windows stuff :) Yeap let's merge! the things i'm don't know/are worried about is publish and signing windows binaries (are they built "correctly" etc?).

wader commented:
🥳
